vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution_unblinded.cc

```diff
@@ -131,176 +114,405 @@ std::string GenerateTokenPayload(
-        reconcile.directions.front().publisher_key,
-        reconcile.type,
-        std::move(list),
-        callback);
+void Unblinded::PreparePublishers(
+    std::vector<ledger::UnblindedToken> list,
+    ledger::ContributionInfoPtr contribution) {
+  if (!contribution) {
     return;
   }
 
-  if (reconcile.type == ledger::RewardsType::AUTO_CONTRIBUTE) {
-    PrepareAutoContribution(viewing_id, std::move(list));
+  if (contribution->type == ledger::RewardsType::AUTO_CONTRIBUTE) {
+    auto publisher_list =
+        PrepareAutoContribution(list, contribution->Clone());
+
+    if (publisher_list.empty()) {
+      ContributionCompleted(
+        ledger::Result::AC_TABLE_EMPTY,
+        std::move(contribution));
+      return;
+    }
+
+    contribution->publishers = std::move(publisher_list);
+
+    ledger_->SaveContributionInfo(
+      contribution->Clone(),
+      std::bind(&Unblinded::OnPrepareAutoContribution,
+          this,
+          _1,
+          contribution->contribution_id));
+    return;
   }
+
+  ProcessTokens(contribution->contribution_id);
 }
 
-bool Unblinded::GetStatisticalVotingWinner(
-    double dart,
-    const ledger::ReconcileDirections& directions,
-    Winners* winners) const {
-  if (!winners) {
-    return false;
+ledger::ContributionPublisherList Unblinded::PrepareAutoContribution(
+    std::vector<ledger::UnblindedToken> list,
+    ledger::ContributionInfoPtr contribution) {
+  if (!contribution || list.size() == 0) {
+    return {};
   }
 
-  double upper = 0.0;
-  for (const auto& item : directions) {
-    upper += item.amount_percent / 100.0;
-    if (upper < dart) {
+  const double total_votes = static_cast<double>(list.size());
+  Winners winners;
+  GetStatisticalVotingWinners(
+      total_votes,
+      contribution->amount,
+      std::move(contribution->publishers),
+      &winners);
+
+  ledger::ContributionPublisherList publisher_list;
+  for (auto & winner : winners) {
+    if (winner.second == 0) {
       continue;
     }
 
-    auto iter = winners->find(item.publisher_key);
+    const std::string publisher_key = winner.first;
+    auto publisher = ledger::ContributionPublisher::New();
+    publisher->contribution_id = contribution->contribution_id;
+    publisher->publisher_key = publisher_key;
+    publisher->total_amount =
+        (winner.second / total_votes) * contribution->amount;
+    publisher->contributed_amount = 0;
+    publisher_list.push_back(std::move(publisher));
+  }
 
-    uint32_t current_value = 0;
-    if (iter != winners->end()) {
-      current_value = winners->at(item.publisher_key);
-      winners->at(item.publisher_key) = current_value + 1;
-    } else {
-      winners->emplace(item.publisher_key, 1);
-    }
+  return publisher_list;
+}
 
-    return true;
+void Unblinded::OnPrepareAutoContribution(
+    const ledger::Result result,
+    const std::string& contribution_id) {
+  if (result != ledger::Result::LEDGER_OK) {
+    SetTimer(&retry_timers_[contribution_id]);
```

Comment:
check does `retry_timers_` contains a `contribution_id` element before accessing it.

Reply:
done